File: apps/androidtv/core/player-contract/src/main/kotlin/dev/spidola/tv/core/playercontract/FakeEngine.kt
// SPDX-FileCopyrightText: 2026 Spidola contributors
// SPDX-License-Identifier: AGPL-3.0-or-later

package dev.spidola.tv.core.playercontract

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update

/**
 * A scriptable in-memory engine for feature-code tests (TECH_SPEC §10: view models are unit-tested
 * against fakes, not real decoders).
 *
 * It ships in the module rather than a test source set because the playback slice's tests, the
 * app's Compose smoke suite, and any future engine's conformance checks all need the same fake —
 * three consumers is exactly when a shared unit is earned. It touches no media framework, so it is
 * the only engine that runs on the JVM without a device.
 *
 * Every state transition is driver-controlled: nothing here is timed, so a test that asserts the
 * loud-fallback path cannot flake on a decode that "usually" takes 40 ms.
 */
class FakeEngine(
    override val id: EngineId = IDENTITY,
) : PlaybackEngine {
    private val _state = MutableStateFlow<PlaybackState>(PlaybackState.Idle)
    override val state: StateFlow<PlaybackState> = _state.asStateFlow()

    private val _tracks = MutableStateFlow(TrackSelection())
    override val tracks: StateFlow<TrackSelection> = _tracks.asStateFlow()

    private val _isSeekable = MutableStateFlow(false)
    override val isSeekable: StateFlow<Boolean> = _isSeekable.asStateFlow()

    /** What [load] was called with — the zap path's assertion surface. */
    val loaded: List<StreamRequest> get() = _loaded.toList()
    private val _loaded = mutableListOf<StreamRequest>()

    var aspect: AspectMode = AspectMode.FIT
        private set

    var isReleased: Boolean = false
        private set

    var playCount: Int = 0
        private set

    var pauseCount: Int = 0
        private set

    val seeks: List<Double> get() = _seeks.toList()
    private val _seeks = mutableListOf<Double>()

    @Composable
    override fun Surface(modifier: Modifier) {
        Box(modifier.background(Color.Black))
    }

    override fun load(request: StreamRequest) {
        _loaded += request
        emit(PlaybackState.Loading)
    }

    override fun play() {
        playCount++
        emit(PlaybackState.Playing)
    }

    override fun pause() {
        pauseCount++
        emit(PlaybackState.Paused)
    }

    override fun seekTo(seconds: Double) {
        if (!_isSeekable.value) return
        _seeks += seconds
    }

    override fun select(track: TrackId) {
        val match = _tracks.value.available.firstOrNull { it.id == track } ?: return
        _tracks.update {
            when (match.kind) {
                TrackKind.AUDIO -> it.copy(selectedAudio = track)
                TrackKind.SUBTITLE -> it.copy(selectedSubtitle = track)
            }
        }
    }

    override fun clearSubtitle() {
        _tracks.update { it.copy(selectedSubtitle = null) }
    }

    override fun setAspect(mode: AspectMode) {
        aspect = mode
    }

    override fun release() {
        isReleased = true
    }

    // region Test driving

    /** Drives the engine to [state], as a real engine's event stream would. */
    fun simulate(state: PlaybackState) = emit(state)

    /** Publishes a track menu, as a real engine does once the stream's tracks are known. */
    fun simulateTracks(
        selection: TrackSelection,
        seekable: Boolean = false,
    ) {
        _tracks.value = selection
        _isSeekable.value = seekable
    }

    // endregion

    private fun emit(state: PlaybackState) {
        if (isReleased) return
        _state.value = state
    }

    companion object {
        /**
         * The engine identity a test resolves to. Distinct from any real engine's key, so a fake
         * can never satisfy a policy assertion that meant a real one.
         */
        val IDENTITY = EngineId("fake")
    }
}

File: apps/tvos/Packages/PlayerContract/Sources/PlayerContract/FakeEngine.swift
// SPDX-FileCopyrightText: 2026 Spidola contributors
// SPDX-License-Identifier: AGPL-3.0-or-later

import SwiftUI

/// A scriptable in-memory engine for feature-code tests (TECH_SPEC §10: view models are unit
/// tested against fakes, not real decoders).
///
/// It ships in the library rather than a test target because the playback slice's tests, the
/// app's UI smoke suite, and any future engine's conformance checks all need the same fake — three
/// consumers is exactly when a shared unit is earned. It touches no media framework, so it is the
/// only engine that runs on a plain build machine.
///
/// Every state transition is driver-controlled: nothing here is timed, so a test that asserts the
/// loud-fallback path cannot flake on a decode that "usually" takes 40 ms.
@MainActor
public final class FakeEngine: PlaybackEngine {
  /// The engine identity a test resolves to. Distinct from any real engine's key, so a fake can
  /// never satisfy a policy assertion that meant a real one.
  public static let identity = EngineID(rawValue: "fake")

  public nonisolated let id: EngineID
  public nonisolated let states: AsyncStream<PlaybackState>

  public private(set) var tracks = TrackSelection()
  public private(set) var isSeekable = false

  /// What `load` was called with — the zap path's assertion surface.
  public private(set) var loaded: [StreamRequest] = []
  public private(set) var aspect: AspectMode = .fit
  public private(set) var isStopped = false
  public private(set) var playCount = 0
  public private(set) var pauseCount = 0
  public private(set) var seeks: [Double] = []

  private let continuation: AsyncStream<PlaybackState>.Continuation
  private var current: PlaybackState = .idle

  public init(id: EngineID = FakeEngine.identity) {
    self.id = id
    // Buffering unbounded: a test that drives five transitions before awaiting the stream must
    // observe all five. A dropping policy would make assertions depend on consumer timing.
    let (stream, continuation) = AsyncStream<PlaybackState>.makeStream(
      bufferingPolicy: .unbounded)
    self.states = stream
    self.continuation = continuation
    continuation.yield(.idle)
  }

  public func makeSurface() -> AnyView {
    AnyView(Color.black)
  }

  public func load(_ request: StreamRequest) {
    loaded.append(request)
    emit(.loading)
  }

  public func play() {
    playCount += 1
    emit(.playing)
  }

  public func pause() {
    pauseCount += 1
    emit(.paused)
  }

  public func seek(toSeconds seconds: Double) {
    guard isSeekable else { return }
    seeks.append(seconds)
  }

  public func select(track: TrackID) {
    guard let match = tracks.available.first(where: { $0.id == track }) else { return }
    switch match.kind {
    case .audio:
      tracks = TrackSelection(
        available: tracks.available, selectedAudio: track,
        selectedSubtitle: tracks.selectedSubtitle)
    case .subtitle:
      tracks = TrackSelection(
        available: tracks.available, selectedAudio: tracks.selectedAudio, selectedSubtitle: track)
    }
  }

  public func clearSubtitle() {
    tracks = TrackSelection(
      available: tracks.available, selectedAudio: tracks.selectedAudio, selectedSubtitle: nil)
  }

  public func setAspect(_ mode: AspectMode) {
    aspect = mode
  }

  public func stop() {
    guard !isStopped else { return }
    isStopped = true
    continuation.finish()
  }

  /// The deinit backstop the contract requires of every engine. The fake holds no decoder, but
  /// keeping the backstop means tests exercise the same lifecycle the real engines have — a test
  /// draining `states` ends when the engine is dropped, exactly as it would against a real one.
  isolated deinit {
    stop()
  }

  // MARK: - Test driving

  /// Drives the engine to `state`, as a real engine's event stream would.
  public func simulate(_ state: PlaybackState) {
    emit(state)
  }

  /// Publishes a track menu, as a real engine does once the stream's tracks are known.
  public func simulateTracks(_ selection: TrackSelection, seekable: Bool = false) {
    tracks = selection
    isSeekable = seekable
  }

  private func emit(_ state: PlaybackState) {
    guard !isStopped else { return }
    current = state
    continuation.yield(state)
    if state.isTerminal { continuation.finish() }
  }

  /// The last state emitted, for tests that assert without draining the stream.
  public var currentState: PlaybackState { current }
}
